rfd/0097-license-expiry.md

```diff
@@ -0,0 +1,147 @@
+
+[RFD 84] specified a license expiry warning to encourage users to renew their
+license. This RFD extends that to disable enterprise features if the license
+still has not been renewed after a grace period.
+
+This RFD proposes to disable SAML and OIDC SSO logins only as they are likely
+the most common reason to use Teleport Enterprise. It does not seem likely that
+other enterprise features are being used without also using the enterprise SSO
+connectors, so it is unnecessary work to disable other features.
+
+[RFD 84]: https://github.com/gravitational/teleport/blob/master/rfd/0084-license-expiration-warnings.md
+
+## Success Criteria
+
+The message shown upon license expiry will warn of upcoming disablement of
+enterprise features.
+
+After a 30-day grace period after licence expiry, the license expired banner
+shown on the web UI and in response to `tsh login`, `tsh status` and `tctl` will
+say that enterprise features have been disabled.
+
+After a 30-day grace period after license expiry, login via SAML or OIDC SSO
+will fail with a license expired error message.
+
+## Scope
+
+This RFD applies to self-hosted Teleport Enterprise. Teleport Cloud is currently
+excluded due to missing license expiration.
+
+## Detail
+
+### UX (CLI and web)
+
+The license checker implemented in [RFD 84] already emits alerts at 90 days
+pre-expiry and at expiry. These alerts are shown when a user runs `tsh login`,
+`tsh status` or any `tctl` command, and are displayed in a banner on the web UI.
+
+The alert for license expiry reads:
+
+    Your Teleport Enterprise Edition license has expired on one or more of your
+    auth servers. Please reach out to licenses@goteleport.com to obtain a new
+    license. Inaction may lead to unplanned outage or degraded performance and
+    support.
+
+The last sentence of that message will be changed to:
+
+    Inaction will lead to enterprise features being disabled and limited
+    support.
+
+An additional alert will be emitted at 30 days post-expiry alerting users that
+enterprise features have been disabled. The banner for this alert will not be
+able to be snoozed or dismissed. The alert will read:
+
+    Your Teleport Enterprise Edition license has expired on one or more of your
+    auth servers. Enterprise features have been disabled. Please reach out to
+    licenses@goteleport.com to obtain a new license.
+
+This follows the existing behaviour as documented in [RFD 84] with the addition
+of the 30 days post-expiry message, and the alteration of the expiry message to
+warn of enterprise features being disabled (Note: it appears it is possible to
+dismiss the expiry message. It appears this is still to be implemented).
+
+### Implementation
+
+The new alert will appear in the response to `ServerWithRoles.GetClusterAlerts`
+as:
+
+```
+{
+  "alerts": [
+    {
+      "kind": "cluster_alert",
+      "version": "v1",
+      "metadata": {
+        "name": "123e4567-e89b-12d3-a456-426614174000",
```

Comment:
If you use a random name, how will you look up this specific alert?

Now that we're starting to use alerts more, maybe we should have a prefix for alerts that teleport creates. Forrest proposed `teleport-<alert-id>`.

I could even envision a system where you register an alert name, some predicate for which the alert should be shown, and the contents of the alert, and a background loop just periodically reconciles it.

```
RegisterInternalAlert("license-expired", func(c *auth.Client) string {
  if !licenseExpired {
    return ""
  }
  if inGracePeriod {
    return "Your license has expired. Inaction will lead to ..."
  }
  
  return "Enterprise features have been disabled..."
})
```